format/bplist/bplist.go

```diff
@@ -0,0 +1,246 @@
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldUTF16("value", int(n))
+	case elementTypeUID:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		d.FieldUBigInt("value", int(n)).Uint64()
+	case elementTypeArray:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		i := uint64(0)
+		d.FieldStructArrayLoop("entries", "entry",
+			func() bool { return i < n },
+			func(d *decode.D) {
+				idx := d.FieldU8("object_index")
+				decodeReference(d, p, idx)
+				i++
+			})
+	case elementTypeSet:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		i := uint64(0)
+		d.FieldStructArrayLoop("entries", "entry",
+			func() bool { return i < n },
+			func(d *decode.D) {
+				idx := d.FieldU8("object_index")
+				decodeReference(d, p, idx)
+				i++
+			})
+	case elementTypeDict:
+		n := decodeSize(d)
+		d.FieldValueU("size", n)
+		i := uint64(0)
+		d.FieldStructArrayLoop("entries", "entry",
+			func() bool { return i < n },
+			func(d *decode.D) {
+				var ki, vi uint64
+				ki = d.FieldU8("key_index")
+				d.SeekRel(int64((n-1)*p.t.objRefSize)*8, func(d *decode.D) {
+					vi = d.FieldU8("value_index")
+				})
+				i++
+				d.FieldStruct("key", func(d *decode.D) {
+					decodeReference(d, p, ki)
+				})
+				d.FieldStruct("value", func(d *decode.D) {
+					decodeReference(d, p, vi)
+				})
+			})
+	default:
+		d.Errorf("unknown type marker: %d", m)
+	}
+}
+
+func decodeReference(d *decode.D, p *plist, idx uint64) {
+	if idx > uint64(len(p.o)) {
+		// prevent a panic
+		d.Errorf("index %d out of bounds for object table size %d", idx, len(p.o))
+		return
+	}
+
+	d.SeekAbs(int64(p.o[idx]*8), func(d *decode.D) {
+		decodeItem(d, p)
+	})
+}
+
+type trailer struct {
+	offTblOffSize    uint64
+	objRefSize       uint64
+	nObjects         uint64
+	topObjectOffset  uint64
+	offsetTableStart uint64
+}
+
+type plist struct {
+	t trailer
+	o []uint64
+}
+
+func bplistDecode(d *decode.D, _ any) any {
+	d.FieldStruct("header", func(d *decode.D) {
+		d.FieldUTF8("magic", 6, d.AssertStr("bplist"))
+		d.FieldUTF8("version", 2, d.AssertStr("00"))
+	})
+
+	p := new(plist)
+
+	d.SeekAbs(d.Len()-32*8, func(d *decode.D) {
+		d.FieldStruct("trailer", func(d *decode.D) {
+			d.FieldU40("padding") // unused
+			d.FieldS8("sort_version")
+			p.t.offTblOffSize = d.FieldU8("offset_table_offset_size", d.AssertURange(1, 8))
+			p.t.objRefSize = d.FieldU8("object_reference_size", d.AssertURange(1, 8))
+			p.t.nObjects = d.FieldU64("object_count")
+			p.t.topObjectOffset = d.FieldU64("top_object_offset")
+			p.t.offsetTableStart = d.FieldU64("offset_table_start")
+		})
+	})
+
+	d.SeekAbs(int64(p.t.offsetTableStart*8), func(d *decode.D) {
+		i := uint64(0)
+		d.FieldArrayLoop("offset_table",
+			func() bool { return i < p.t.nObjects },
+			func(d *decode.D) {
+				off := d.FieldU("element", 8*int(p.t.offTblOffSize))
+				p.o = append(p.o, off)
+				i++
+			},
+		)
+	})
+
+	d.SeekAbs(int64(p.o[0] * 8))
+
+	d.FieldStruct("objects",
+		func(d *decode.D) {
+			decodeItem(d, p)
+		})
```

Comment:
Could this somehow be `decodeReference(d, p, 0)`? looks similar but maybe confusing?

Reply:
I think that works, I prefer it over the seek operation.